### PR TITLE
*: refactor the RestrictedSQLExecutor interface (#22579)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #22579 to release-5.0-rc
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/22687
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/22687:release-5.0-rc-ea6ccf82e934
```

---

### What problem does this PR solve?

Problem Summary:

Improve security for our code.

### What is changed and how it works?

The initial definition of `RestrictedSQLExecutor` looks like this:

```
type RestrictedSQLExecutor interface {
	// ExecRestrictedSQL run sql statement in ctx with some restriction.
	ExecRestrictedSQL(sql string) ([]chunk.Row, []*ast.ResultField, error)
}
```

It's not good for security reasons, the usage pattern looks like this:

```
sql  = fmt.Sprintf('select xxx from ..%s ', varFromOutside)
stmt, err = ExecRestrictedSQL(sql) 
...
```

Later on, we add more methods to the interface which IMO, ugly:

```
	// ExecRestrictedSQLWithContext run sql statement in ctx with some restriction.
	ExecRestrictedSQLWithContext(ctx context.Context, sql string, opts ...OptionFuncAlias) ([]chunk.Row, []*ast.ResultField, error)
	// ExecRestrictedSQLWithSnapshot run sql statement in ctx with some restriction and with snapshot.
	// If current session sets the snapshot timestamp, then execute with this snapshot timestamp.
	// Otherwise, execute with the current transaction start timestamp if the transaction is valid.
	ExecRestrictedSQLWithSnapshot(sql string) ([]chunk.Row, []*ast.ResultField, error)
```

I propose to use this one as its new definition:

```
type RestrictedSQLExecutor interface {
	ParseWithParams(ctx context.Context, sql string, args ...interface{}) (ast.StmtNode, error)
	ExecRestrictedStmt(ctx context.Context, stmt ast.StmtNode, opts ...OptionFuncAlias) ([]chunk.Row, []*ast.ResultField, error)
}
```

This interface is secure, and easy to use:

```
stmt  = exec.ParseWithParams('select xxx from ..%? ', varFromOutside)
res, err = ExecRestrictedSQL(stmt) 
...
```

What's Changed:

`ParseWithParams()` now returns `ast.StmtNode` rather than `[]ast.StmtNode`, multiple statements are not allowed here.

Update the "util", "store", and "domain" packages,  ensure they use the new (secure) API.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- No code

Side effects

- Breaking backward compatibility
API change

### Release note 

- No release note

### CODE REVIEW

**Comment by bb7133** [comment]
Please try to resolve the conflicts @tiancaiamao

File: store/tikv/gcworker/gc_worker.go [hunk 1]
```diff
@@ -1756,14 +1756,12 @@ func (w *GCWorker) loadValueFromSysTable(key string) (string, error) {
 	se := createSession(w.store)
 	defer se.Close()
 	rs, err := se.ExecuteInternal(ctx, `SELECT HIGH_PRIORITY (variable_value) FROM mysql.tidb WHERE variable_name=%? FOR UPDATE`, key)
-	if len(rs) > 0 {
-		defer terror.Call(rs[0].Close)
-	}
+	defer terror.Call(rs.Close)
```
Comment (xhebox) [inline store/tikv/gcworker/gc_worker.go:1759]: This line should be moved after the `err != nil` check

**Review by xhebox — COMMENTED** [review]
Rest LGTM

**Review by xhebox — COMMENTED** [review]
LGTM

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**